ballerina-to-openapi/src/main/java/io/ballerina/openapi/service/mapper/interceptor/InterceptorPipeline.java

```diff
@@ -0,0 +1,276 @@
+ * the interceptor pipeline and extracting the effective return types from the pipeline.
+ *
+ * @since 1.9.0
+ */
+public class InterceptorPipeline {
+
+    private Interceptor initReqInterceptor = null;
+    private Interceptor initResInterceptor = null;
+    private final SemanticModel semanticModel;
+
+    private InterceptorPipeline(SemanticModel semanticModel) {
+        this.semanticModel = semanticModel;
+    }
+
+    public static InterceptorPipeline build(ServiceDeclarationNode serviceDefinition, SemanticModel semanticModel,
+                                            ModuleMemberVisitor moduleMemberVisitor) {
+        List<Interceptor> interceptors = buildInterceptors(serviceDefinition, semanticModel, moduleMemberVisitor);
+
+        InterceptorPipeline pipeline = new InterceptorPipeline(semanticModel);
+        if (!interceptors.isEmpty()) {
+            Interceptor prevInterceptor = interceptors.get(0);
+            if (prevInterceptor.getType().equals(InterceptorType.REQUEST)) {
+                pipeline.initReqInterceptor = prevInterceptor;
+            }
+            if (prevInterceptor instanceof ResponseInterceptor) {
+                pipeline.initResInterceptor = prevInterceptor;
+            }
+            for (int i = 1; i < interceptors.size(); i++) {
+                prevInterceptor.setNextInReqPath(interceptors.get(i));
+                interceptors.get(i).setNextInResPath(prevInterceptor);
+                prevInterceptor = interceptors.get(i);
+                if (prevInterceptor.getType().equals(InterceptorType.REQUEST) &&
+                        Objects.isNull(pipeline.initReqInterceptor)) {
+                    pipeline.initReqInterceptor = prevInterceptor;
+                } else if (prevInterceptor instanceof ResponseInterceptor) {
+                    pipeline.initResInterceptor = prevInterceptor;
+                }
+            }
+        }
+        return pipeline;
+    }
+
+    private static List<Interceptor> buildInterceptors(ServiceDeclarationNode serviceDefinition,
+                                                       SemanticModel semanticModel,
+                                                       ModuleMemberVisitor moduleMemberVisitor) {
+        Optional<Symbol> optServiceSymbol = semanticModel.symbol(serviceDefinition);
+        if (optServiceSymbol.isEmpty() ||
+                !(optServiceSymbol.get() instanceof ServiceDeclarationSymbol serviceSymbol)) {
+            return new ArrayList<>();
+        }
+
+        Optional<TypeSymbol> optInterceptorReturn = serviceSymbol.methods().get("createInterceptors").
+                typeDescriptor().returnTypeDescriptor();
+        if (optInterceptorReturn.isEmpty()) {
+            return new ArrayList<>();
+        }
+        if (!(optInterceptorReturn.get() instanceof TupleTypeSymbol interceptorTupleType)) {
+            // Print a warning since OAS can not extract interceptors from a
```

Comment:
Did not understand this comment. You mean when we return an empty arraylist, we catch it somewhere and print the warning?

Reply:
Ah, this need to be updated. Thanks for pointing this out.
Ideally we will add a diagnostic and return an empty array. When the interceptor array is empty we will fallback to the default response mapping

Reply:
Fixed via [9e19c0a](https://github.com/ballerina-platform/openapi-tools/pull/1614/commits/9e19c0a9c57916d7d6ba8c0de9a4676afd5a2a1b)